File: elixir/triangle/triangle.exs
defmodule Triangle do
  @type kind :: :equilateral | :isosceles | :scalene

  @doc """
  Return the kind of triangle of a triangle with 'a', 'b' and 'c' as lengths.
  """
  @spec kind(number, number, number) :: { :ok, kind } | { :error, String.t }
  def kind(a, a, a) when a > 0 do { :ok, :equilateral } end
  def kind(a, b, c) when a > 0 and b > 0 and c > 0 do
    cond do
      degnerate?(a,b,c) -> { :error, "side lengths violate triangle inequality"  }
      degnerate?(a,c,b) -> { :error, "side lengths violate triangle inequality"  }
      degnerate?(c,b,a) -> { :error, "side lengths violate triangle inequality"  }
      a == b -> { :ok, :isosceles }
      a == c -> { :ok, :isosceles }
      b == c -> { :ok, :isosceles }
      true -> { :ok, :scalene }
    end
  end
  def kind(_, _, _) do { :error, "all side lengths must be positive"  } end

  defp degnerate?(a, b, c) when a + b > c do false end
  defp degnerate?(a, b, c) when a + b < c do true end
  defp degnerate?(a, b, c) when a == b do true end

end

File: go/triangle/triangle.go
// Package triangle determines the type of triangle based on side lengths
package triangle

import (
	"math"
)

const testVersion = 3

type Kind int

const NaT Kind = 0 // not a triangle
const Equ Kind = 1 // equilateral
const Iso Kind = 2 // isosceles
const Sca Kind = 3 // scalene

func KindFromSides(a, b, c float64) Kind {
	switch {
	case invalidSide(a) || invalidSide(b) || invalidSide(c):
		return NaT
	case a+b < c || a+c < b || b+c < a:
		return NaT
	case a == b && a == c:
		return Equ
	case a == b || a == c || b == c:
		return Iso
	}

	return Sca
}

func invalidSide(n float64) bool {
	return n == 0 || math.IsNaN(n) || math.IsInf(n, 0)
}
